Joss Paper (#503)
sunpy/ndcube: joss_paper/paper.md

---
title: 'ndcube: Manipulating N-dimensional Astronomical Data in Python'
tags:
  - Python
  - astronomy
authors:
  - name: Daniel F. Ryan^[co-first author] # note this makes a footnote saying 'co-first author'
    orcid: 0000-0001-8661-3825
    affiliation: 1 # (Multiple affiliations must be quoted)
  - name: Stuart Mumford^[co-first author] # note this makes a footnote saying 'co-first author'
    affiliation: 2
  - name: Author with no affiliation^[corresponding author]
    affiliation: 3
affiliations:
 - name: University of Applied Sciences Northwest Switzerland
   index: 1

>  - name: University of Sheffield

Reviewer: ```suggestion
 - name: Aperio Software Ltd
```

   index: 2
date: 11 December 2021
bibliography: paper.bib

---

# Summary

ndcube is a free, open-source, community-developed Python package whose purpose
is to faciliate the manipulation and visualization of astronomical data.

> It does this by linking the data and coordinates in single data objects.

Reviewer: "single data object**s**" reads weird.

Author: unified data objects?

These objects can be manipulated via array-like slicing operations which modify
the data and coordinates simultaneously.
They also allow coordinate transformations to be performed with reference to the
shape of the data and enable coordinate-aware visualizations.
This data-coordinate coupling allows users to analyze their data more

> easily and accurately, thus helping to boost their scientific output.

Reviewer: I think we should probably mention, and reference, Astropy's APE 14 here, that in reality is what makes us unique among the wider scientific python ecosystem. https://zenodo.org/record/1188875

This leaves more time for developing a greater range of tools for the community
and/or enables part-time developers to devote more effort to other aspects of their jobs,
e.g. scientific analysis.

A network benefit of ndcube is that it standardizes the APIs for handling N-D data.
Broad adoption throughout astronomy and heliophysics would help scientists to more easily
work with data from different sub-fields.
This can help facilitate synergies between new combinations of data,
foster inter-field collaborations, and promote scientific innovation.

# Statement of Need

N-dimensional data sets are common in all areas of science and beyond.
For example, a series of images taken sequentially with a CCD camera can be stored
as a single 3-D array with two spatial axes and one temporal axis.
The value in each array-element can represent the reading in a pixel at a given time.
In astronomy, the relationship between the array element and the location and time
in the Universe being observed is often represented by the World Coordinate System (WCS) framework.
WCS’s ability to handle many different physical types (e.g. spatial, temporal, spectral, etc.)

> and projections (e.g. RA and Dec., helioprojective latitude and longitude, etc.)

Reviewer: We should mention this is a *functional* framework where the transformation is modelled and then evaluated on demand, vs lookup tables.

make it a succinct, standardized and powerful way to relate array axes to the physical
properties they represent.
Due of the prevalence of N-D data and the importance of coordinate transformations,
there exist mature Python packages for handling them.
However, they are not suited to treating data and coordinates in a combined way.
This is the purpose of ndcube.

The fundamental reason to opt for ndcube over its most similar package, xarray,
is to harness the astronomy-specific World Coordinate System (WCS).
The data model of xarray centers on the requirements and conventions of the geosciences.
Although similar to those of astronomy in conception, they are sufficiently different
in construction to cause significant friction.
Moreover, utilizing the astropy WCS infrastructure enables us to directly read the most

> common file format in astronomy, FITS.

Reviewer: Do we need this?!

Author: No!  I've removed the whole paragraph locally.

# Data Objects: The Pillars of ndcube

At the time of writing (ndcube v2.0), ndcube provides three primary data objects for

> manipulating astronomical data: NDCube, NDCubeSequence, and NDCollection.

Reviewer: ```suggestion
manipulating astronomical data: `NDCube`, `NDCubeSequence`, and `NDCollection`.
```

Each each cube is referenced by its name, making NDCollection similar to a Python dictionary.
However it also enables axes to be marked as "aligned" which facilitates
simultaneous slicing of the collection's components along those aligned axes.
In the above example, we could mark the spatial axes of the spectral image cube and
doppler map as "aligned" and then extract a spatial sub-region by slicing the NDCollection
rather than the two data objects separately.
Thus NDCollection also enables easier and more reliable manipulation of N-D astronomical
data sets.

# Citations

Citations to entries in paper.bib should be in
[rMarkdown](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html)
format.

If you want to cite a software repository URL (e.g. something on GitHub without a preferred
citation) then you can do it with the example BibTeX entry below for @fidgit.

For a quick reference, the following citation commands can be used:
- `@author:2001`  ->  "Author et al. (2001)"
- `[@author:2001]` -> "(Author et al., 2001)"
- `[@author1:2001; @author2:2001]` -> "(Author1 et al., 2001; Author2 et al., 2002)"

# Acknowledgements

> We acknowledge support for ndcube from NASA's Heliophysics Data Environment Enhancement
> program and the Daniel K. Inouye Solar Telescope.
> We also acknowledge the SunPy and Python in Heliophysics communities for their support.

Author: ```suggestion
The ndcube developers would like to thank the Google Summer of Code program,
NASA's Heliophysics Data Environment Enhancement program,
and the Daniel K. Inouye Solar Telescope for their support.
They would also like to thank the SunPy and Python in Heliophysics communities for their support.
```
